### DISCUSSION TITLE
how to share spdlog in libraries and main function?

### DISCUSSION THREAD

**Discussion by gakki2019** [gabime/spdlog#3171] [Q&A] [answered]
There are several libraries and one process main function. All libraries and main record to same log file or stdout.
Is there any sample? (Linux platform)

I try to put the code in singleton,  like logger.hpp
```
class logger {
public:
    static spdlog::logger *get_instance() {
        static logger instance;
        return instance.m_logger.get();
    }
    private:
    logger() noexcept;
    ~logger();
    logger(const logger &) = delete;
    logger &operator=(const logger &) = delete;

private:
    std::shared_ptr<spdlog::logger> m_logger;
};

logger::logger() noexcept {
    auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
    console_sink->set_level(spdlog::level::trace);
    console_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%l]%v");

    auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>("log.txt", 1024*1024, 5);
    file_sink->set_level(spdlog::level::trace);
    file_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%l]%v");

    m_logger = std::make_shared<spdlog::logger>(spdlog::logger{"mylogger", {console_sink, file_sink}});
    //spdlog::register_logger(m_logger);
    m_logger->set_level(spdlog::level::debug);
    m_logger->flush_on(spdlog::level::err);
}

logger::~logger() {
    //spdlog::shutdown();
   printf("logger  ---\n");
}
```

libmyso1.so/libmyso2.so/libmyso3.so are almost same like:
```
extern "C" __attribute((visibility("default"))) void fnSo1() {
    auto log = logger::get_instance();

    log->info("This is an info message from libso1.");
    log->error("This is an error message from libso1.");
}
```

main.cpp
```
int main() {
auto log = logger::get_instance();

    log->info("This is an info message from main.");
    log->error("This is an error message from main.");

    fnSo1();
    fnSo2();
    fnSo3();
}

```
as for main process Makefile
CLIB                      += -L$(MY_LIB_PATH) -lmyso1 -lmyso2 -lmyso3

after lanuch, report double free 
![image](https://github.com/user-attachments/assets/496c1329-14cf-4583-b6a5-5dfbfa5c676c)

**Comment by tt4g** [comment]
Example: https://github.com/gabime/spdlog/wiki/How-to-use-spdlog-in-DLLs

By the way, is the double free occurring in spdlog?

**Reply by tt4g** [reply]
Perhaps the `logger` function body is actually in `logger.hpp`?
If so, then `logger` may have copies in all libraries and applications.

A `static` function has been copied into the translation unit, and there could be many instances of `logger`.
Some C++ versions add an implicit `inline` keyword: [Storage class specifiers - cppreference.com](https://en.cppreference.com/w/cpp/language/storage_duration))

Move the `logger` function body to `logger.cpp` and create one new library to share `logger`.

See also: [Why can you have the method definition inside the header file in C++ when in C you cannot? - Software Engineering Stack Exchange](https://softwareengineering.stackexchange.com/questions/56215/why-can-you-have-the-method-definition-inside-the-header-file-in-c-when-in-c-y)

**Reply by gakki2019** [reply]
Thanks for reply.

By the way, is the double free occurring in spdlog?
[A] yes.

logger::logger() {
   ...
 }
 
 function body is in logger.cpp,not hpp

**Reply by tt4g** [reply]
> function body is in logger.cpp,not hpp

Hmmm, it is unlikely that double free will occur as long as the smart pointer is used.

Could you also tell me why you determined that spdlog was the cause of the double free?

When gdb detects a double free, the backtrace shows the function name that caused it.
`__do_global_dtors_aux()` and `__cxa_finalize()` are functions to call destructors of global variables. Since the name `/opt/lib/libaxcl_so2.so` appears beside `__do_global_dtors_aux()`, it is assumed that a double free has occurred during the call to the destructor of a global variable in this library.
If spdlog were responsible, the spdlog function name would appear after it.
However, I cannot find both the spdlog function name and your provided source code function name in the backtrace.

**Reply by tt4g** [reply]
Maybe related #2113

**Reply by gakki2019** [reply]
I found a clue. I think it is still related to "extern" of UNIX library as descibled in [ https://github.com/gabime/spdlog/discussions/2198](url)
In my code,  logger.hpp.   _**static std::string m_log_path**_;  is declared as class variable as shown as below:

```
 #pragma once
#include "spdlog/spdlog.h"
#include <string>

#define LOG_PATH 

class logger {
public:
    static spdlog::logger *get_instance() {
        static logger instance;
        return instance.m_logger.get();
    }

private:
    logger() noexcept;
    ~logger();
    logger(const logger &) = delete;
    logger &operator=(const logger &) = delete;

private:
    std::shared_ptr<spdlog::logger> m_logger;

#ifdef LOG_PATH
    static std::string m_log_path;
#endif
};
```

**If remove  _static std::string m_log_path;,_ program works well.**
I don't fully understand the library memory system of UNIX, and if you known, could you tell me?

**Reply by tt4g** [reply, marked as answer]
In C++, declaring a `static` variable in a class declaration creates a copy of the same variable in all modules that include the header file.
Do not declare `static` variables in the class body.
It is safe to declare `static` variables in class functions written in `cpp` files.

- [c\+\+ \- Why do we need to define static variables of a class, but we don't define static variables when they are inside functions? \- Stack Overflow](https://stackoverflow.com/questions/68450957/why-do-we-need-to-define-static-variables-of-a-class-but-we-dont-define-static)
- [static members \- cppreference\.com](https://en.cppreference.com/w/cpp/language/static)

**Reply by gakki2019** [reply]
logger.cpp
```
#include "logger.hpp"
#include <spdlog/sinks/rotating_file_sink.h>
#include <spdlog/sinks/stdout_color_sinks.h>
#include <stdio.h>
#ifdef LOG_PATH
std::string logger::m_log_path = "logs/axcl_logs.txt";
#endif

logger::logger() noexcept {
    m_logger = spdlog::get("mylogger");
    if (!m_logger) {
        printf("create logger\n");
        auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
        console_sink->set_level(spdlog::level::trace);
        console_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%l]%v");

        auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>("log.txt", 1024 * 1024, 5);
        file_sink->set_level(spdlog::level::trace);
        file_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%l]%v");

        m_logger = std::make_shared<spdlog::logger>(spdlog::logger{"mylogger", {console_sink, file_sink}});
        spdlog::register_logger(m_logger);
        m_logger->set_level(spdlog::level::debug);
        m_logger->flush_on(spdlog::level::err);
    } else {
        printf("logger has been created\n");
    }
}

logger::~logger() {
    // spdlog::shutdown();
}
```

**Reply by gakki2019** [reply]
thanks a lot.
